rust/lance/src/dataset.rs

```diff
@@ -1116,6 +1184,27 @@ impl Dataset {
         Ok(versions)
     }
 
+    /// Get all tags.
+    pub async fn tags(&self) -> Result<HashMap<String, TagContents>> {
+        let mut tags = HashMap::<String, TagContents>::new();
+
+        let tag_names = self
+            .object_store()
+            .read_dir(base_tags_path(&self.base))
+            .await?;
+
+        for n in tag_names.iter() {
+            let tag_name = n.strip_suffix(".json").unwrap();
+            let tag_path = tag_path(&self.base, tag_name);
+            tags.insert(
+                tag_name.to_string(),
+                TagContents::from_path(&tag_path, self.object_store()).await?,
+            );
+        }
```

Comment:
Can support some concucurrency in loading the tags by using a stream and `buffer_unordered`. Also, for safety, might want to filter out any non-json suffixes instead of panicking when those occur.
```suggestion
        futures::stream::iter(tag_names.iter())
            .filter_map(|name| name.strip_suffix(".json").map(|n| n.to_string()))
            .map(|tag_name| {
                let tag_path = tag_path(&self.base, &tag_name);
                async move {
                    let contents = TagContents::from_path(&tag_path, self.object_store()).await?;
                    Ok((tag_name, contents))
                }
            )
            .buffer_unordered(10)
            .for_each(|(tag_name, contents)| {
                tags.insert((tag_name, contents))
            });
        }
```

Reply:
I committed this without testing locally - it doesn't seem to compile yet. I'll work in this change later.

Reply:
I've addressed this issue in 01db9ae. It required a bit of wrangling to include error handling and make it compile.